src/backend/distributed/executor/executor.c

```diff
@@ -779,6 +783,57 @@ LockPartitionsForDistributedPlan(DistributedPlan *distributedPlan)
 }
 
 
+/*
+ * AcquireExecutorShardLocks acquires advisory lock on shard IDs to prevent
+ * unsafe concurrent modifications of shards.
+ *
+ * We prevent concurrent modifications of shards in two cases:
+ * 1. Any non-commutative writes to a replicated table
+ * 2. Multi-shard writes that are executed in parallel
+ *
+ * The first case ensures we do not apply updates in different orders on
+ * different replicas (e.g. of a reference table), which could lead the
+ * replicas to diverge.
+ *
+ * The second case prevents deadlocks due to out-of-order execution.
+ *
+ * We do not take executor shard locks for utility commands such as
+ * TRUNCATE because the table locks already prevent concurrent access.
+ */
+static void
+AcquireExecutorShardLocks(DistributedExecution *execution)
+{
+	CmdType operation = execution->operation;
+	List *taskList = execution->tasksToExecute;
+
+	if (!(operation == CMD_INSERT || operation == CMD_UPDATE || operation == CMD_DELETE))
```

Comment:
I think `SELECT FOR UPDATE` on reference tables should go through `AcquireExecutorShardLock `.  So, we should probably not have this check.

But then `AcquireExecutorMultiShardLocks ` should not be called for `SELECTs` because it doesn't have any check for it.

Reply:
We could maybe just ignore SELECT FOR UPDATE in this PR, and have it via #2768